pypose/function/geometry.py

```diff
@@ -159,3 +175,210 @@ def reprojerr(points, pixels, intrinsics, extrinsics=None):
+        >>> pp.knn(pc1, pc2)
+        torch.return_types.topk(
+        values=tensor([[2.0000],
+                [1.0000],
+                [1.4142],
+                [1.4142],
+                [0.0000],
+                [4.2426]]),
+        indices=tensor([[3],
+                [0],
+                [0],
+                [2],
+                [0],
+                [1]]))
+        >>> pp.knn(pc1, pc2, k=2, ord=2)
+        torch.return_types.topk(
+        values=tensor([[2.0000, 4.5826],
+                [1.0000, 4.5826],
+                [1.4142, 5.0990],
+                [1.4142, 4.0000],
+                [0.0000, 4.2426],
+                [4.2426, 5.0000]]),
+        indices=tensor([[3, 2],
+                [0, 2],
+                [0, 2],
+                [2, 0],
+                [0, 2],
+                [1, 2]]))
+        >>> pp.knn(pc1, pc2, k=2, ord=2).values
+        tensor([[2.0000, 4.5826],
+                [1.0000, 4.5826],
+                [1.4142, 5.0990],
+                [1.4142, 4.0000],
+                [0.0000, 4.2426],
+                [4.2426, 5.0000]])
+    '''
+    diff = pc1.unsqueeze(-2) - pc2.unsqueeze(-3)
+    dist = torch.linalg.norm(diff, dim=dim, ord=ord)
+    neighbors = dist.topk(k, dim=dim, largest=largest, sorted=sorted)
+    return neighbors
+
+
+def svdtf(pc1, pc2):
+    r'''
+    Computes the rigid transformation ( :math:`SE(3)` ) between two sets of associated
+    points using Singular Value Decomposition (SVD).
+
+    Args:
+        pc1 (``torch.Tensor``): The coordinates of the first set of points.
+            The shape has to be (..., N, 3).
+        pc2 (``torch.Tensor``): The coordinates of the second set of points.
+            The shape has to be (..., N, 3).
+
+    Returns:
+        ``LieTensor``: The rigid transformation matrix in ``SE3Type``  that
+        minimizes the mean squared error between the input point sets.
+
+    Warning:
+        The number of points N has to be the same for both point clouds.
+
+    Example:
+        >>> import torch, pypose as pp
+        >>> pc1 = torch.tensor([[0., 0., 0.],
+        ...                     [1., 0., 0.],
+        ...                     [0., 1., 0.]])
+        >>> pc2 = torch.tensor([[1., 1., 1.],
+        ...                     [2., 1., 1.],
+        ...                     [1., 2., 1.]])
+        >>> pp.svdtf(pc1, pc2)
+        SE3Type LieTensor:
+        LieTensor([1., 1., 1., 0., 0., 0., 1.])
+    '''
+    assert pc1.size(-2) == pc2.size(-2), "The number of the points N has to be the same."
+    pc1ctn = pc1.mean(dim=-2, keepdim=True)
+    pc2ctn = pc2.mean(dim=-2, keepdim=True)
+    pc1t = pc1 - pc1ctn
+    pc2t = pc2 - pc2ctn
+    M = bvv(pc2t, pc1t).sum(dim=-3)
+    U, S, Vh = torch.linalg.svd(M)
+    R = U @ Vh
+    mask = (R.det() + 1).abs() < 1e-6
+    R[mask] = - R[mask]
+    t = pc2ctn.mT - R @ pc1ctn.mT
+    T = torch.cat((R, t), dim=-1)
+    return mat2SE3(T, check=False)
+
+
+def posediff(ref, est, aggregate=False, mode=1):
```

Comment:
it seems that this function is not used in any functions for now. Here is my concern on the API:
- The mode 0 contains quaternion, which should be a LieTensor, as users may want to use its properties for further computation.
- How will `posediff` be used is still not clear, I suggest not to expose to users for now, as once release we will need to maintain this API and this will pose issues if we find a better API in the future.

Reply:
Agree. This function is currently used in test_icp and potentially will be used in other tests. The initial idea is to calculate the error between two transformations, where the translational error is easy to define, but the rotational ones are a bit tricky (like mentioned in https://github.com/pypose/pypose/issues/187). This function is not clearly structured with limited functionality at current stage. I think we can make it an internal function for now, and have another pull request dedicate to it.